apps/passport-server/src/services/poapService.ts

```diff
@@ -0,0 +1,253 @@
+import { getHash } from "@pcd/passport-crypto";
+import { SerializedPCD } from "@pcd/pcd-types";
+import { ZKEdDSAEventTicketPCDPackage } from "@pcd/zk-eddsa-event-ticket-pcd";
+import AsyncLock from "async-lock";
+import { fetchDevconnectPretixTicketByTicketId } from "../database/queries/devconnect_pretix_tickets/fetchDevconnectPretixTicket";
+import {
+  claimNewPoapUrl,
+  getExistingClaimUrlByTicketId
+} from "../database/queries/poap";
+import { ApplicationContext } from "../types";
+import { logger } from "../util/logger";
+import { getServerErrorUrl } from "../util/util";
+import { RollbarService } from "./rollbarService";
+import { traced } from "./telemetryService";
+
+// Set up an async-lock to prevent concurrency issues when two threads
+// are simultaneously calling `getDevconnectPoapClaimUrl()`.
+const lock = new AsyncLock();
+const POAP_CLAIM_LOCK_KEY = "poap-claim-lock-key";
+
+const DEVCONNECT_COWORK_SPACE_EVENT_ID = "a1c822c4-60bd-11ee-8732-763dbf30819c";
+
+// All valid Cowork products that can claim a POAP. This excludes add-on products, like the Turkish Towel.
+const DEVCONNECT_COWORK_SPACE_VALID_PRODUCT_IDS = [
+  "67687bda-986f-11ee-abf3-126a2f5f3c5c",
+  "67689552-986f-11ee-abf3-126a2f5f3c5c",
+  "6768a2e0-986f-11ee-abf3-126a2f5f3c5c",
+  "6768af1a-986f-11ee-abf3-126a2f5f3c5c",
+  "6768c81a-986f-11ee-abf3-126a2f5f3c5c",
+  "6768d44a-986f-11ee-abf3-126a2f5f3c5c",
+  "6768e21e-986f-11ee-abf3-126a2f5f3c5c",
+  "6768ecf0-986f-11ee-abf3-126a2f5f3c5c",
+  "6768f7cc-986f-11ee-abf3-126a2f5f3c5c",
+  "67690410-986f-11ee-abf3-126a2f5f3c5c",
+  "67690e92-986f-11ee-abf3-126a2f5f3c5c",
+  "67691888-986f-11ee-abf3-126a2f5f3c5c",
+  "67692468-986f-11ee-abf3-126a2f5f3c5c",
+  "676932d2-986f-11ee-abf3-126a2f5f3c5c",
+  "67694cfe-986f-11ee-abf3-126a2f5f3c5c",
+  "676961d0-986f-11ee-abf3-126a2f5f3c5c"
+];
+
+/**
+ * Responsible for issuing POAP (poap.xyz) mint links to users
+ * who have attended a certain event, e.g. Devconnect.
+ */
+export class PoapService {
+  private readonly context: ApplicationContext;
+  private readonly rollbarService: RollbarService | null;
+
+  public constructor(
+    context: ApplicationContext,
+    rollbarService: RollbarService | null
+  ) {
+    this.context = context;
+    this.rollbarService = rollbarService;
+  }
+
+  /**
+   * Validates that a serialized ZKEdDSAEventTicketPCD is a valid
+   * Devconnect Cowork ticket that has been checked in, and returns
+   * the ID of that ticket.
+   *
+   * This function throws an error in the case that the PCD is not
+   * valid; for example, here are a few invalid cases
+   *  1. Wrong PCD type
+   *  2. Wrong EdDSA public key
+   *  3. PCD proof is invalid
+   *  4. Tjcket does not exist
+   *  5. Ticket has not been checked in
+   *  6. Event of ticket is not Cowork space
+   *  7. Invalid product for claiming a poap, e.g. EF Towel
+   */
+  private async validateDevconnectTicket(
+    serializedPCD: string
+  ): Promise<string> {
```

Comment:
for the `logger` logs, it's worth logging some extra info that is specific to the particular invocation of this function, whether that's the pcd, or some data derived from it like the ticket id. otherwise the logs tell us that something happened but not exactly what happened.